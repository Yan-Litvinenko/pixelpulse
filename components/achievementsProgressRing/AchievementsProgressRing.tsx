'use client';

import React from 'react';
import useClient from '@/hooks/useClient';

export default function AchievementsProgressRing({ percent }: { percent: number }): React.JSX.Element {
    const isClient = useClient();
    const square: number = 148;
    const border: number = 14;
    const CYX: number = square / 2;
    const radius: number = square / 2 - border * 2;
    const circumference: number = 2 * Math.PI * radius;

    return isClient ? (
        <svg
            className={'progress_ring'}
            width={`${square}rem`}
            height={`${square}rem`}
            viewBox={`0 0 ${square} ${square}`}
            data-testid="svg"
        >
            <circle
                style={{
                    strokeDashoffset: `${circumference - (percent / 100) * circumference}`,
                    strokeDasharray: `${circumference} ${circumference}`,
                }}
                className={'progress_ring__circle'}
                stroke="current-color"
                strokeWidth={border}
                cy={CYX}
                cx={CYX}
                r={radius}
                fill="transparent"
                data-testid="circle"
            />
        </svg>
    ) : (
        <></>
    );
}
