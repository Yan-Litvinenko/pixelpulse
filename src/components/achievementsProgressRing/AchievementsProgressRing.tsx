import React from 'react';
import styles from './AchievementsProgressRing.module.scss';
import type { IAchievementsProgressRing } from '../../interfaces/interface.achievements';

const AchievementsProgressRing = ({ percent }: IAchievementsProgressRing): React.JSX.Element => {
    const square: number = 148;
    const border: number = 14;
    const CYX: number = square / 2;
    const radius: number = square / 2 - border * 2;
    const circumference: number = 2 * Math.PI * radius;

    return (
        <svg
            className={styles.ring}
            width={square}
            height={square}
            viewBox={`0 0 ${square} ${square}`}
            data-testid="svg"
        >
            <circle
                style={{
                    strokeDashoffset: `${circumference - (percent / 100) * circumference}`,
                    strokeDasharray: `${circumference} ${circumference}`,
                }}
                className={styles.ring__circle}
                stroke="current-color"
                strokeWidth={border}
                cy={CYX}
                cx={CYX}
                r={radius}
                fill="transparent"
                data-testid="circle"
            />
        </svg>
    );
};

export { AchievementsProgressRing };
