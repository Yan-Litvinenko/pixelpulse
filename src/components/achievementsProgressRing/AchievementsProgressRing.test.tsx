import React from 'react';
import { AchievementsProgressRing } from './AchievementsProgressRing';
import { render, screen } from '@testing-library/react';

describe('AchievementsProgressRing component', (): void => {
    test('Renders correctly with given percent', (): void => {
        const percent = 75;
        const square = 148;
        const border = 14;
        const CYX = square / 2;
        const radius = square / 2 - border * 2;
        const circumference = 2 * Math.PI * radius;

        render(<AchievementsProgressRing percent={percent} />);

        const svg = screen.getByTestId('svg');
        const circle = screen.getByTestId('circle');

        expect(svg).toBeInTheDocument();
        expect(circle).toBeInTheDocument();

        expect(circle.getAttribute('stroke-width')).toBe(`${border}`);
        expect(circle.getAttribute('cy')).toBe(`${CYX}`);
        expect(circle.getAttribute('cx')).toBe(`${CYX}`);
        expect(circle.getAttribute('r')).toBe(`${radius}`);

        expect(circle).toHaveStyle(`stroke-dashoffset: ${circumference - (percent / 100) * circumference}`);
        expect(circle).toHaveStyle(`stroke-dasharray: ${circumference} ${circumference}`);

        expect(svg).toMatchSnapshot();
    });

    test('Renders correctly with 0 percent', (): void => {
        const percent = 0;
        const square = 148;
        const border = 14;
        const radius = square / 2 - border * 2;
        const circumference = 2 * Math.PI * radius;
        const dashoffset = `stroke-dashoffset: ${circumference - (percent / 100) * circumference}`;

        render(<AchievementsProgressRing percent={percent} />);

        const circle = screen.getByTestId('circle');

        expect(circle).toBeInTheDocument();
        expect(circle).toHaveStyle(dashoffset);
    });

    test('Renders correctly with 100 percent', (): void => {
        const percent = 100;

        render(<AchievementsProgressRing percent={percent} />);

        const circle = screen.getByTestId('circle');

        expect(circle).toBeInTheDocument();
        expect(circle).toHaveStyle('stroke-dashoffset: 0');
    });
});
