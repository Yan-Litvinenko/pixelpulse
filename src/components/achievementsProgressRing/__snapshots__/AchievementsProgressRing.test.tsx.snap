// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AchievementsProgressRing component Renders correctly with given percent 1`] = `
<svg
  data-testid="svg"
  height="148"
  viewBox="0 0 148 148"
  width="148"
>
  <circle
    cx="74"
    cy="74"
    data-testid="circle"
    fill="transparent"
    r="46"
    stroke="current-color"
    stroke-width="14"
    style="stroke-dashoffset: 72.25663103256522; stroke-dasharray: 289.02652413026095 289.02652413026095;"
  />
</svg>
`;
